File: Premium/Zigzag-Iterator.cpp
class ZigzagIterator {
private:
	int idx1, idx2;
	vector<int> &vec1, &vec2;
public:
	ZigzagIterator(vector<int>& v1, vector<int>& v2) : vec1(v1), vec2(v2) {
		idx1 = idx2 = 0;
	}

	int next() {
		if (idx1 == vec1.size()) {
			return vec2[idx2++];
		}
		else if (idx2 == vec2.size()) {
			return vec1[idx1++];
		}
		else {
			if (idx1 > idx2) {
				return vec2[idx2++];
			}
			else {
				return vec1[idx1++];
			}
		}
	}

	bool hasNext() {
		return idx1 != vec1.size() || idx2 != vec2.size();
	}
};

/**
 * Your ZigzagIterator object will be instantiated and called as such:
 * ZigzagIterator i(v1, v2);
 * while (i.hasNext()) cout << i.next();
 */

File: Java/Zigzag-Iterator.java
public class ZigzagIterator {
    public int x1, x2;
    public ArrayList<Integer> arr1;
    public ArrayList<Integer> arr2;
    public ZigzagIterator(List<Integer> v1, List<Integer> v2) {
        arr1 = new ArrayList<Integer>(v1);
        arr2 = new ArrayList<Integer>(v2);
        x1 = 0;
        x2 = 0;
    }

    public int next() {
        if (x1 == arr1.size()) {
            return arr2.get(x2 ++);
        } else if (x2 == arr2.size()) {
            return arr1.get(x1 ++);
        } else {
            if (x1 > x2) {
                return arr2.get(x2 ++);
            } else {
                return arr1.get(x1 ++);
            }
        }
    }

    public boolean hasNext() {
        return x1 != arr1.size() || x2 != arr2.size();
    }
}
